File: tests/support/test_home.py
"""Give each standalone binary driver disposable application homes."""
import os
import pathlib
import tempfile

# Retained until interpreter exit; subprocesses finish before fixture cleanup.
_home = None


def isolate(checks_enabled=False):
    global _home
    import json

    if "HOME" in os.environ:
        original_home = pathlib.Path(os.environ["HOME"])
        os.environ.setdefault("RUSTUP_HOME", str(original_home / ".rustup"))
        os.environ.setdefault("CARGO_HOME", str(original_home / ".cargo"))
    _home = tempfile.TemporaryDirectory(prefix="rustrace-driver-home-")
    root = pathlib.Path(_home.name)
    for key, directory in [("HOME", "home"), ("XDG_CONFIG_HOME", "config"),
                           ("XDG_STATE_HOME", "state")]:
        path = root / directory
        path.mkdir()
        os.environ[key] = str(path)
    state = root / "state/rustrace/update-state.json"
    state.parent.mkdir()
    state.write_text(json.dumps({"schema_version": 1, "checks_enabled": checks_enabled,
                                "last_attempt": None, "last_success": None,
                                "next_eligible": None, "latest": None}))

File: tests/support/test_home.rs
//! Fixture-owned application homes. Keep the guard alive while children run.
use std::{
    fs,
    path::PathBuf,
    process::Command,
    sync::atomic::{AtomicU64, Ordering},
};

pub struct TestHome {
    pub root: PathBuf,
}

impl TestHome {
    pub fn new(checks_enabled: bool) -> Self {
        static NEXT: AtomicU64 = AtomicU64::new(0);
        let root = std::env::temp_dir().join(format!(
            "rustrace-test-home-{}-{}",
            std::process::id(),
            NEXT.fetch_add(1, Ordering::Relaxed)
        ));
        fs::create_dir_all(root.join("state/rustrace")).unwrap();
        fs::create_dir_all(root.join("config")).unwrap();
        fs::create_dir_all(root.join("home")).unwrap();
        fs::write(
            root.join("state/rustrace/update-state.json"),
            serde_json::to_vec(&rustrace::update::UpdateState {
                checks_enabled,
                ..rustrace::update::UpdateState::default()
            })
            .unwrap(),
        )
        .unwrap();
        Self { root }
    }

    pub fn command(&self, program: impl AsRef<std::ffi::OsStr>) -> Command {
        let mut command = Command::new(program);
        // Preserve installed tooling when changing HOME, without writing there.
        if let Some(home) = std::env::var_os("HOME") {
            let home = PathBuf::from(home);
            for (key, fallback) in [("RUSTUP_HOME", ".rustup"), ("CARGO_HOME", ".cargo")] {
                command.env(
                    key,
                    std::env::var_os(key).unwrap_or_else(|| home.join(fallback).into_os_string()),
                );
            }
        }
        command
            .env("HOME", self.root.join("home"))
            .env("XDG_CONFIG_HOME", self.root.join("config"))
            .env("XDG_STATE_HOME", self.root.join("state"));
        command
    }
}

impl Drop for TestHome {
    fn drop(&mut self) {
        let _ = fs::remove_dir_all(&self.root);
    }
}
